### PR TITLE
AppProvider Cleanup

### PR DESCRIPTION
**Pull request by kvendrik**
### WHY are these changes introduced?
fixes https://github.com/Shopify/polaris-react-deprecated/issues/2361
moved from https://github.com/Shopify/polaris-react-deprecated/pull/2362

### WHAT is this pull request doing?
Separates all utilities and their dependencies and moves them into their own `utilities/` folder.

### CODE REVIEW

**Comment by danrosenthal** [comment]
@kvendrik what's the status on this PR?

**Comment by kvendrik** [comment]
@danrosenthal it was on hold until the v3 release. I've pinged @BPScott yesterday about finding someone to finish this up if he would like to see it merged soon as I'm currently busy with other projects.

**Comment by dpersing** [comment]
Friendly neighborhood staleness checker 🍞! Got a bit of time to check @BPScott?

**Comment by kvendrik** [comment]
This PR is waiting for review by @BPScott, pinged him about it just now 😄

File: src/components/AppProvider/utilities/Link/Link.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import {LinkLikeComponent} from '../UnstyledLink';
```
Comment (BPScott) [inline src/components/AppProvider/utilities/Link/Link.ts]: Keeps as a relative path for the moment

Comment (BPScott) [reply src/components/AppProvider/utilities/Link/Link.ts]: (Gad this should be a comment on the diff line below, you changed this from a relative to absolute link)

**Comment by danrosenthal** [comment]
> Volunteering @danrosenthal for this as Tim said Dan is probably more familiar with this)

👀

File: src/components/AppProvider/AppProvider.tsx [hunk 2]
```diff
@@ -77,7 +46,7 @@ export default class AppProvider extends React.Component<Props> {
     apiKey,
     shopOrigin,
     forceRedirect,
-  }: Props) {
+  }: AppProviderProps) {
     if (
       i18n !== this.props.i18n ||
       linkComponent !== this.props.linkComponent ||
```
Comment (danrosenthal) [inline src/components/AppProvider/AppProvider.tsx]: 👍

File: src/components/AppProvider/index.ts [hunk 3]
```diff
@@ -1,19 +1,22 @@
...
-  SCROLL_LOCKING_ATTRIBUTE,
-  SCROLL_LOCKING_CUSTOM_PROPERTY,
-} from './ScrollLockManager';
-
-export default AppProvider;
+  AppProviderProps as Props,
+  Context,
+  polarisAppProviderContextTypes,
+} from './types';
+export {default} from './AppProvider';
```
Comment (danrosenthal) [inline src/components/AppProvider/index.ts]: 👍

File: src/components/AppProvider/tests/AppProvider.test.tsx [hunk 4]
```diff
@@ -1,11 +1,8 @@
...
-
-import {createAppProviderContext} from '../utils';
 import {createThemeContext} from '../../ThemeProvider';
+import {StickyManager, createAppProviderContext} from '../utilities';
 import {polarisAppProviderContextTypes} from '../types';
-import StickyManager from '../StickyManager';
-
 import AppProvider from '../AppProvider';
 
 describe('<AppProvider />', () => {
```
Comment (danrosenthal) [inline src/components/AppProvider/tests/AppProvider.test.tsx]: 👍

File: src/components/AppProvider/types.ts [hunk 5]
```diff
@@ -1,50 +1,48 @@
...
-export interface CreateAppProviderContext extends AppProviderProps {
-  stickyManager?: StickyManager;
-  scrollLockManager?: ScrollLockManager;
-  subscribe?(callback: () => void): void;
-  unsubscribe?(callback: () => void): void;
-}
+export const polarisAppProviderContextTypes: ValidationMap<any> = {
+  polaris: PropTypes.any,
+  ...polarisTheme,
+};
```
Comment (danrosenthal) [inline src/components/AppProvider/types.ts]: 👍

File: src/components/AppProvider/utilities/Intl/Intl.ts [hunk 6]
```diff
@@ -1,16 +1,13 @@
...
   TranslationDictionary,
   PrimitiveReplacementDictionary,
   ComplexReplacementDictionary,
 } from './types';
-
-import {translate} from './utils';
+import translate from './translate';
 
 export interface ChangeTranslation {
   (translation: TranslationDictionary): void;
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/Intl/Intl.ts]: 👍

File: src/components/AppProvider/utilities/Intl/index.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export {default} from './Intl';
+export {default as translate} from './translate';
+export {
+  TranslationDictionary,
+  PrimitiveReplacementDictionary,
+  ComplexReplacementDictionary,
+} from './types';
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/Intl/index.ts]: 👍

File: src/components/AppProvider/utilities/Intl/translate.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            .join(', ')}`,
+        );
+      }
+
+      return replacements[replacement] as string;
+    });
+  }
+
+  return text;
+}
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/Intl/translate.ts]: 👍

File: src/components/AppProvider/utilities/Intl/types.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  [key: string]: string | TranslationDictionary;
+}
+
+export interface PrimitiveReplacementDictionary {
+  [key: string]: string | number;
+}
+
+export interface ComplexReplacementDictionary {
+  [key: string]: string | number | React.ReactNode;
+}
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/Intl/types.ts]: 👍

File: src/components/AppProvider/utilities/createAppProviderContext/createAppProviderContext.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+      intl,
+      link,
+      stickyManager: stickyManager || new StickyManager(),
+      scrollLockManager: scrollLockManager || new ScrollLockManager(),
+      subscribe,
+      unsubscribe,
+      appBridge,
+    },
+  };
+}
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/createAppProviderContext/createAppProviderContext.ts]: 👍

File: src/components/AppProvider/utilities/createAppProviderContext/tests/createAppProviderContext.test.tsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+          apiKey,
+          forceRedirect: undefined,
+          shopOrigin: undefined,
+        },
+      },
+    };
+
+    expect(context).toEqual(mockContext);
+  });
+});
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/createAppProviderContext/tests/createAppProviderContext.test.tsx]: 🙏

File: src/components/AppProvider/utilities/Intl/tests/translate.test.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      );
+    });
+
+    it('throws an error for a missing replacement', () => {
+      expect(() =>
+        translateWithReplacements('foo {next}', {notNext: 'bar'}),
+      ).toThrow();
+    });
+  });
+});
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/Intl/tests/translate.test.ts]: 👍

File: src/components/AppProvider/utilities/createPolarisContext/tests/createPolarisContext.test.tsx [hunk 13]
```diff
@@ -1,124 +1,10 @@
...
-  });
-});
+import Intl from '../../Intl';
+import {StickyManager} from '../../withSticky';
+import Link from '../../Link';
+import ScrollLockManager from '../../ScrollLockManager';
+import createPolarisContext from '../createPolarisContext';
 
 describe('createPolarisContext()', () => {
   it('returns the right context without arguments', () => {
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/createPolarisContext/tests/createPolarisContext.test.tsx]: 👍

File: src/components/AppProvider/utilities/withAppProvider/withAppProvider.tsx [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    const FinalComponent = hoistStatics(
+      WithProvider,
+      WrappedComponent as React.ComponentClass<any>,
+    );
+
+    return FinalComponent as React.ComponentClass<OwnProps> & C;
+  };
+}
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/withAppProvider/withAppProvider.tsx]: This needs tests badly. I'll make a few issues following my review for adding test coverage.

File: src/components/AppProvider/utilities/withSticky/withSticky.tsx [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+      }
+    }
+
+    const FinalComponent = hoistStatics(
+      WithStickyManager,
+      WrappedComponent as React.ComponentClass<any>,
+    );
+    return FinalComponent as React.ComponentClass<any> & C;
+  };
+}
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/withSticky/withSticky.tsx]: Similarly this is in need of tests. Will open an issue.

File: src/components/AppProvider/utils/index.tsx [hunk 16]
```diff
@@ -1,248 +0,0 @@
...
-
-  const appProvider = appProviderContext
-    ? createAppProviderContext(appProviderContext)
-    : createAppProviderContext();
-  const theme = themeContext
-    ? createThemeContext(themeContext)
-    : createThemeContext();
-
-  return {...appProvider, ...theme};
-}
```
Comment (danrosenthal) [inline src/components/AppProvider/utils/index.tsx]: 👋

File: src/components/AppProvider/utilities/createPolarisContext/createPolarisContext.ts [hunk 17]
```diff
@@ -17,7 +18,7 @@ export interface CreatePolarisContext extends AppProviderProps {
 
 export const polarisAppProviderContextTypes: ValidationMap<any> = {
   polaris: PropTypes.any,
-  easdk: PropTypes.any,
```
Comment (danrosenthal) [inline src/components/AppProvider/utilities/createPolarisContext/createPolarisContext.ts]: I guess we forgot this little guy.

**Comment by danrosenthal** [comment]
@BPScott you good with me going ahead and merging this?

**Comment by kvendrik** [comment]
> Personally I would probably put all the utilities files in the root of that folder (e.g. utilities/createAppProviderContext.tsx instead of utilities/createAppProviderContext/index.ts and utilities/createAppProviderContext/createAppProviderContext.tsx) but that's just personal preference.

The reason they're in seperate folders is so that we can start adding tests for specific utilities e.g. `utilities/createAppProviderContext/tests/createAppProviderContext.test.ts` 🙂 .

**Comment by kvendrik** [comment]
> The only downside of this PR is it highlights the need for tests in the following:

This PR is part of https://github.com/Shopify/polaris-react-deprecated/issues/2361#issue-366511462 of which the goal is to get the test coverage up to 80%. 😄

File: src/components/AppProvider/utilities/createPolarisContext/createPolarisContext.ts (hunk 17 above)
Comment (kvendrik) [reply src/components/AppProvider/utilities/createPolarisContext/createPolarisContext.ts]: Think I manually removed it.

**Comment by danrosenthal** [comment]
> of which the goal is to get the test coverage up to 80%+

Awesome! That will be huge. Thanks @kvendrik

**Comment by BPScott** [comment]
Thanks so much for the review @danrosenthal. If you're happy I'm happy to merge

**Pull request merged**